// relaen/dist/core/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transaction = void 0;
const logger_1 = require("./logger");
const tipmanager_1 = require("./message/tipmanager");
const threadlocal_1 = require("./threadlocal");
const transactionfactory_1 = require("./transactionfactory");
/**
 * 事务基类
 */
class Transaction {
    /**
     * 构造器
     * @param conn -  真实连接对象
     */
    constructor(conn) {
        this.conn = conn;
        //如果无threadid，则新建
        this.threadId = threadlocal_1.RelaenThreadLocal.getThreadId() || threadlocal_1.RelaenThreadLocal.newThreadId();
    }
    /**
     * 设置当前事务
     * @param isolationLevel - 事务隔离级
     */
    async setIsolationLevel(isolationLevel) {
        this.isolation = isolationLevel;
    }
    /**
     * 事务开始
     */
    async begin() {
        logger_1.Logger.log(tipmanager_1.RelaenTipManager.getTip('txbegin'));
    }
    /**
     * 事务提交,继承类需要重载
     */
    async commit() {
        //从事务管理器移除
        transactionfactory_1.TransactionFactory.remove();
        logger_1.Logger.log(tipmanager_1.RelaenTipManager.getTip('txcommit'));
    }
    /**
     * 事务回滚,继承类需要重载
     */
    async rollback() {
        //从事务管理器移除
        transactionfactory_1.TransactionFactory.remove();
        logger_1.Logger.log(tipmanager_1.RelaenTipManager.getTip('txrollback'));
    }
}
exports.Transaction = Transaction;
